File: 04-stack/00853-Car_Fleet/00853-car_fleet.go
// 853: Car Fleet
// https://leetcode.com/problems/car-fleet/

package main

import (
	"fmt"
	"sort"
)


type car struct {
	p int
	s int
}

// SOLUTION
func carFleet(target int, position []int, speed []int) int {
	cars := []car{}
	n := len(position)
	for i:=0; i<n; i++ {
		cars = append(cars, car{position[i], speed[i]})
	}

	sort.Slice(cars, func(i, j int) bool {
		return cars[i].p < cars[j].p
	})

	s := []float64{}
	for i:=0; i<n; i++ {
		time := float64((target - cars[i].p) / cars[i].s)
		for len(s)!=0 && time >= s[len(s)-1] {s = s[:len(s)-1]}
		s = append(s, time)
	}

	return len(s);
}


func main() {
	// INPUT
	target := 12
    position := []int{10,8,0,5,3}
    speed := []int{2,4,1,1,3}

    // OUTPUT
    result := carFleet(target, position, speed)
    fmt.Println(result)
}

File: 04-stack/00853-Car_Fleet/00853-car_fleet.py
# 853: Car Fleet
# https://leetcode.com/problems/car-fleet/


import operator


class Solution:
    class car:
        def __init__(self, p: int, s: int):
            self.p = p
            self.s = s

    # SOLUTION
    def carFleet(self, target: int, position: list[int], speed: list[int]) -> int:
        cars: list[car] = []
        n = len(position)
        for i in range(n):
            cars.append(self.car(position[i], speed[i]))

        cars.sort(key=operator.attrgetter('p'))

        s: list[float] = []
        for i in range(n):
            time = (target - cars[i].p) / cars[i].s
            while len(s)!=0 and time >= s[-1]: s.pop()
            s.append(time)

        return len(s)



if __name__ == "__main__":
    o = Solution()

    # INPUT
    target: int = 12
    position: list[int] = [10,8,0,5,3]
    speed: list[int] = [2,4,1,1,3]

    # OUTPUT
    result = o.carFleet(target, position, speed)
    print(result)